### ISSUE TITLE
TypeError: Failed to convert object of type <class 'list'> to Tensor. Contents: [1, 1, Dimension(None)]. Consider casting elements to a supported type.

### ISSUE THREAD

**Issue by nd7141** [tensorflow/models#6245]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux
- **TensorFlow version (use command below)**: 1.12

I'm trying to reshape some of my data with Dataset API. In particular, I have created a dataset like this: 

```
dataset = tf.data.TFRecordDataset(filenames)
dataset = dataset.map(parse).filter(filter_empty).batch(batch_size)
```

After this if I print my shape within map function I get the following:

```
def print_shape(x):
    print(x.shape)
    
dataset.map(print_shape)
```

`>>> (?, ?, ?)`

So the shape of the batch is not clear and hence if I want to reshape some of my variables by accessing them via `x.shape`, it fails with an error: 
`TypeError: Failed to convert object of type <class 'list'> to Tensor. Contents: [1, 1, Dimension(None)]. Consider casting elements to a supported type.`

How can I access shape within dataset creation so that I cam reshape some of my variables?

**Comment by nd7141** [comment]
Another detail: I use eager mode.

**Comment by nd7141** [comment]
Solved via using `tf.shape` instead.

**Comment by Shruthi-Sampathkumar** [comment]
> Solved via using `tf.shape` instead.

How did you use tf.shape to solve this issue?

**Comment by sachin5nowal** [comment]
I have been facing similar issues ,try using  tf.expand_dims.
`x=tf.expand_dims(x, -1)`
This helped me sorting the issue.

**Comment by Liusandian** [comment]
> Solved via using `tf.shape` instead.

How did you use tf.shape to solve this issue? could you please give more details about the solution?

**Comment by wookayin** [comment]
@Liusandian @Shruthi-Sampathkumar 

Let's assume `x` is a Tensor of shape [10, ?, ?]. It can be a tensor with a undetermined size when it is TF 1.x or running non-eagerly on TF 2.x (e.g. using `tf.function` + autograph)

`x.shape` will give you `[10, None, None]`. You cannot reshape it because you do not know the dimension statically. For example, some tensor manipulations `tf.reshape(x, [-1, x.shape[1], x.shape[0]])` will fail.

However, `h = tf.shape(x)[1]; w = tf.shape(x)[2]` will let `h`, `w` be symbolic or graph-mode tensors (integer) that will contain a dimension of `x`. The value will be determined runtime. In such a case, `tf.reshape(x, [-1, w, h])` will produce a (symbolic) tensor of shape `[?, ?, ?]` (still unknown) whose tensor shape will be known on runtime.

TL;DR) `tf.shape()` gives a `tf.Tensor` of integer dtype which is not `None`.

**Comment by QinHsiu** [comment]
w is a matrix
input_dim is 190 
x is a outputs of a network with shape [None,190]
when I use the following command:
tf.tensordot(tf.reshape(x, [-1, 1, input_dim]), w, 1)
It raises TypeError(f"Failed to convert elements of {values} to Tensor. "
Could anyone help me with this problem?

**Comment by Liusandian** [comment]
这是来自QQ邮箱的假期自动回复邮件。   谢谢你的来信,我会尽快给你回复。
